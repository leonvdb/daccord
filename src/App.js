import React, { Component } from 'react';
import { BrowserRouter, Route } from 'react-router-dom';

import Landing from './components/Landing';
import Header from './components/layout/Header';
import Footer from './components/layout/Footer';
import Poll from './components/poll/Poll';
import CreatePoll from './components/CreatePoll';

import './App.css';

class App extends Component {
  render() {
    return (
      <BrowserRouter>
        <div className="App">
          <Header />
          <Route exact path="/" component={Landing} />
          <Route exact path="/poll" component={Poll} />
          <Route exact path="/create" component={CreatePoll} />
          <Footer />
        </div>
      </BrowserRouter>
    );
  }
}

export default App;
