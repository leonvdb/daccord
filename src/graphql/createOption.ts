import gql from 'graphql-tag';

export const createOption = gql`
mutation CreateOption(
        $pollId: ID!,
        $userId: ID!,
        $title: String!,
        $description: String
    ) {
        createOption(
            pollId: $pollId,
            userId: $userId,
            title: $title,
            description: $description
        ){
            title
            description
        }
    }

`