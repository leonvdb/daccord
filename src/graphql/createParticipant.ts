import gql from 'graphql-tag';

export const CREATE_PARTICIPANT = gql`
mutation CreateParticipant($pollId: ID!, $email: String!, $pseudonym: String!){
    createParticipant(pollId: $pollId, email: $email, pseudonym: $pseudonym){
        user{
            id
            email
        }
        token
        pseudonym
    }
}
`