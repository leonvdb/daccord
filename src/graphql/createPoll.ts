import gql from 'graphql-tag';

export const CREATE_POLL = gql`

mutation CreatePoll($title: String!, $userEmail: String!){
    createPoll(title: $title, userEmail: $userEmail){
        user{
            id
            email
        }
        poll{
            refId
        }
        token
    }
}
`