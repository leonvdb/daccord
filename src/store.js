import { createStore, compose, applyMiddleware } from 'redux';
import rootReducer from './reducers';
import thunk from 'redux-thunk';

const initialState = {};

const middlewares = [thunk];

function createApplicationStore() {
    // Production Store
    if (process.env.NODE_ENV === 'production') {
        return createStore(rootReducer, initialState, compose(
            applyMiddleware(...middlewares),
        ));
    }

    // Development store
    if (window.__REDUX_DEVTOOLS_EXTENSION__) {
        return createStore(rootReducer, initialState, compose(
            applyMiddleware(...middlewares),
            window.__REDUX_DEVTOOLS_EXTENSION__ &&
            window.__REDUX_DEVTOOLS_EXTENSION__()
        ));
    }
    // Devtools missing
    return createStore(rootReducer, initialState, compose(
        applyMiddleware(...middlewares),
    ));
}

const store = createApplicationStore()

export default store;

