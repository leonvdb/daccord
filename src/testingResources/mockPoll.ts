import { IPollQuery } from '../interfaces';

const mockPoll: IPollQuery = {
    title: "Test Poll",
    description: "Description of the Test Poll",
    refId: "TestRefId",
    creatorPseudonym: "Test Creator",
    creator: {
        id: "TestCreatorId",
        email: "creator@example.com",
        name: "Test Creator"
    },
    options: [{
        title: "Test Option",
        description: "Test Option Description",
        creator: {
            id: "TestCreatorId",
            email: "creator@example.com",
            name: "Test Creator"
        },
        refId: "TestOptionRefId",
        userRating: 8,
        result: {
            totalOpposition: 8,
            participationInPercent: 100,
            agreementInPercent: 20
        }
    }],
    participants: [{
        pseudonym: "Test Participant",
        user: {
            id: "TestParticipantId",
            email: "participant@example.com",
            name: "Test Participant"
        }
    }]
}

export default mockPoll;