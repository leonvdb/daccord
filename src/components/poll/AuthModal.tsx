import * as React from 'react'
import { Modal, ModalHeader, ModalBody } from 'reactstrap';
import { connect } from 'react-redux';
import { Store } from '../../reducers';
import { IUser, IPollQuery } from '../../interfaces';
import TextInputGroup from '../layout/TextInputGroup';
import validateEmail from '../../utilities/validateEmail';
import { Dispatch } from 'redux';
import { CREATE_PARTICIPANT } from '../../graphql/createParticipant';
import { Mutation, compose, withApollo } from 'react-apollo';
import { setAuthTokenAndUser } from '../../actions/authActions';
import { SEND_AUTH_LINK } from '../../graphql/sendAuthLink';
import DefaultClient from 'apollo-boost';
import { setPseudonym } from '../../actions/userActions';


interface Props extends PropsFromState, PropsFromDispatch {
    poll: IPollQuery
    isOpen: boolean
    renderButton: boolean
    client: DefaultClient<any>
}
interface State {
    isOpen: boolean,
    showParticipantError: boolean
    showLinkSent: boolean
    name: string,
    email: string,
    errors: Errors
}
interface Errors {
    name?: string,
    email?: string
}

class AuthModal extends React.Component<Props> {
    state: State = {
        isOpen: this.props.isOpen,
        showParticipantError: false,
        showLinkSent: false,
        name: '',
        email: '',
        errors: {}
    }

    onChange = (e: React.ChangeEvent<any>) => {
        const propertyName = e.target.name
        const value = e.target.value
        this.setState(prevState => {
            const newState = { ...prevState };
            newState[propertyName] = value
            return newState
        })
    };

    onSubmit = (e: React.FormEvent<HTMLFormElement>, mutation: any) => {
        e.preventDefault();
        const { name, email } = this.state
        const { poll } = this.props

        // Form validation
        const errors: Errors = {}
        if (name.length <= 0) { errors.name = 'Please enter a name' }
        if (!validateEmail(email)) { errors.email = 'Please enter a valid email address.' }
        if (Object.keys(errors).length > 0) {
            this.setState({
                errors
            })
            return;
        }

        mutation({variables: {pollId: poll.refId, email, pseudonym: name}})

    }

    toggle = () => {
        this.setState({ isOpen: !this.state.isOpen })
    }

    backFromError = () => {
        this.setState({
            showParticipantError: false,
            showLinkSent: false
        })
    }

    resendLink = () => {
        this.props.client.mutate({mutation: SEND_AUTH_LINK, variables: {pollId: this.props.poll.refId, email: this.state.email}})
        this.setState({ showLinkSent: true })
    }

    render() {

        const { isOpen, name, email, errors, showParticipantError, showLinkSent } = this.state
        const { poll, renderButton } = this.props

        let modal;
        if (!showParticipantError) {
            modal = <Modal placement="right" isOpen={isOpen} target="Modal" toggle={this.toggle}>
                <ModalHeader toggle={this.toggle}>
                    Become a participant of "{poll.title}"
                    </ModalHeader>
                <ModalBody>
                    <Mutation 
                    mutation={CREATE_PARTICIPANT}
                    update={// tslint:disable-next-line jsx-no-lambda
                        (cache, { data: { createParticipant}}) => {
                            if (createParticipant.token) {
                                this.props.setAuthTokenAndUser(createParticipant.user, createParticipant.token)
                                this.props.setPseudonym(createParticipant.pseudonym)
                            } else {
                                this.setState({showParticipantError: true})
                            }
                        }}
                    >
                        {(CREATE_PARTICIPANT) => (
                    <form onSubmit={// tslint:disable-next-line jsx-no-lambda
                        (e) => {this.onSubmit(e, CREATE_PARTICIPANT)}}>
                        <TextInputGroup
                            label="Name"
                            name="name"
                            placeholder="Enter Name"
                            value={name}
                            onChange={this.onChange}
                            error={errors.name}
                            classNames="w-75"
                        />
                        <TextInputGroup
                            label="Email"
                            name="email"
                            placeholder="Enter Email"
                            value={email}
                            onChange={this.onChange}
                            error={errors.email}
                            classNames="w-75 mb-5"
                        />
                        <button className="btn btn-secondary mx-auto btn-block w-50 mt-5" type="submit">Continue</button>
                    </form>
                        )} 
                    </Mutation>
                    <p className="text-center my-2">or</p>
                    <button className="btn btn-outline-info btn-block w-50 mx-auto">Sign in</button>
                </ModalBody>
            </Modal>
        } else {
            if (!showLinkSent) {
                modal =
                    <Modal placement="right" isOpen={isOpen} target="Modal" toggle={this.toggle}>
                        <ModalHeader toggle={this.toggle}>
                            <i className="fas fa-arrow-left mr-5" onClick={this.backFromError} style={{ cursor: 'pointer' }} />
                            Become a participant of "{poll.title}"
                    </ModalHeader>
                        <ModalBody>
                            <div className="alert alert-danger">
                                Seems like you are already participating
                </div>
                            <button onClick={this.resendLink} className="btn btn-link btn-block mx-auto">Request new access link</button>
                        </ModalBody>
                    </Modal>
            } else {
                modal = <Modal placement="right" isOpen={isOpen} target="Modal" toggle={this.toggle}>
                    <ModalHeader toggle={this.toggle}>
                        <i className="fas fa-arrow-left mr-5" onClick={this.backFromError} style={{ cursor: 'pointer' }} />
                        Become a participant of "{poll.title}"
                    </ModalHeader>
                    <ModalBody>
                        <div className="alert alert-success">
                            Your new access Link has been sent, please check your email.
                </div>
                    </ModalBody>
                </Modal>
            }
        }

        return (
            <div className="col-sm-6 col-md-4 col-lg-3 mb-4 d-flex justify-content-center align-items-center">
                {renderButton && <button
                    id="Modal"
                    onClick={this.toggle}
                    className="btn btn-outline-success">
                    <i className="fas fa-hand-peace mr-1" />Participate
                </button>}
                {modal}
            </div>
        )
    }
}

const mapStateToProps = (state: Store) => ({
    user: state.user.user,
});

interface PropsFromDispatch {
    setAuthTokenAndUser: (user: IUser, jwt: string) => void
    setPseudonym: (pseudonym: string) => void
}

const mapDispatchToProps = (dispatch: Dispatch<any>): PropsFromDispatch => {
    return {
        setAuthTokenAndUser: ( user: IUser, jwt: string) => dispatch(setAuthTokenAndUser(user, jwt)),
        setPseudonym: (pseudonym: string) => dispatch(setPseudonym(pseudonym))
    }

}

interface PropsFromState {
    user: IUser
}

export default compose(
    withApollo,
    connect(mapStateToProps, mapDispatchToProps)
)(AuthModal) as any;