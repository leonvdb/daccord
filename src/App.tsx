import * as React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter, Route, Switch } from 'react-router-dom';
import { I18nextProvider } from "react-i18next";
import ApolloClient from 'apollo-boost';
import { ApolloProvider } from 'react-apollo';
import store from './store';

import CreatePoll from './components/CreatePoll';
import Landing from './components/Landing';
import Header from './components/layout/Header';
import NotFound from './components/NotFound';
import Poll from './components/poll';

import './App.css';
import i18n from './i18n';
import setAuthToken from './utilities/setAuthToken';

export const client = new ApolloClient({
  uri: 'http://localhost:5000/graphql',
  request: async operation => {
    operation.setContext({
      headers: {
        Authorization: localStorage.getItem('jwtToken')
      }
    })
  }
})

if (localStorage.jwtToken) {
  setAuthToken(localStorage.jwtToken)
}

class App extends React.Component {
  public render() {
    return (
      <ApolloProvider client={client}>
        <I18nextProvider i18n={i18n}>
          <Provider store={store}>
            <BrowserRouter>
              <div className="App">
                <Header />
                <Switch>
                  <Route exact={true} path="/" component={Landing} />
                  <Route exact={true} path="/poll/:poll_id" component={Poll} />
                  <Route name="pollNavRoute" path="/poll/:poll_id/:pollNavRoute" component={Poll} />
                  <Route exact={true} path="/create" component={CreatePoll} />
                  <Route component={NotFound} />
                </Switch>
              </div>
            </BrowserRouter>
          </Provider>
        </I18nextProvider>
      </ApolloProvider>
    );
  }
}

export default App;
