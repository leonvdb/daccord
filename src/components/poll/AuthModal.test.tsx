import React from 'react';
import AuthModal from './AuthModal'
import { render, fireEvent, cleanup, waitForElement } from 'react-testing-library';
import { ApolloProvider } from 'react-apollo';
import { client } from '../../App';
import { Provider } from 'react-redux';
import store from '../../store';
import mockPoll from '../../testingResources/mockPoll'
import { CREATE_PARTICIPANT } from '../../graphql/createParticipant';
import { MockedProvider } from 'react-apollo/test-utils';

const originalError = console.error
beforeAll(() => {
    console.error = (...args: any) => {
        if (/Warning.*not wrapped in act/.test(args[0])) {
            return
        }
        originalError.call(console, ...args)
    }
})

afterAll(() => {
    console.error = originalError
})

afterEach(cleanup);

const createMock = (email: string, name: string, returnToken: boolean) => ([
    {
        request: {
            query: CREATE_PARTICIPANT,
            variables: {
                pollId: mockPoll.refId,
                email,
                pseudonym: name
            }
        },
        result: {
            data: {
                createParticipant: {
                    user: {
                        id: 'mockId',
                        email
                    },
                    token: returnToken ? 'newParticipantToken' : '',
                    pseudonym: name
                }
            }
        },
    },
])

test('<AuthModal /> to show Modal on Click', () => {
    const { getByTestId, queryByTestId } = render(
        <Provider store={store}>
            <ApolloProvider client={client}>
                <AuthModal poll={mockPoll} renderButton={true} />
            </ApolloProvider>
        </Provider>
    );
    expect(queryByTestId('auth-modal')).toBeFalsy();
    const participateButton = getByTestId('participate-button');
    fireEvent.click(participateButton);
    expect(getByTestId('auth-modal'));
    expect(getByTestId('auth-modal')).toMatchSnapshot();
})

test('<AuthModal /> rejects invalid input', () => {
    const { getByTestId, getByPlaceholderText } = render(
        <Provider store={store}>
            <ApolloProvider client={client}>
                <AuthModal poll={mockPoll} isOpen={true} />
            </ApolloProvider>
        </Provider>
    );

    expect(getByTestId('auth-modal'));
    const emailInput = (getByPlaceholderText("Enter Email") as HTMLInputElement)
    fireEvent.change(emailInput, {
        target: {
            value: 'invalid Email'
        }
    })
    expect(emailInput.value).toBe('invalid Email')
    fireEvent.click(getByTestId('submit-button'))
    expect(getByTestId('error-message'))
})

test('<AuthModa /> raises error for already existing participants', async () => {
    const { getByPlaceholderText, getByTestId } = render(
        <Provider store={store}>
            <MockedProvider mocks={createMock(mockPoll.creator.email, mockPoll.creator.name, false)} addTypename={false}>
                <AuthModal poll={mockPoll} isOpen={true} />
            </MockedProvider>
        </Provider>
    );

    const emailInput = (getByPlaceholderText("Enter Email") as HTMLInputElement)
    const nameInput = (getByPlaceholderText("Enter Name") as HTMLInputElement)
    fireEvent.change(emailInput, {
        target: {
            value: mockPoll.creator.email
        }
    })
    fireEvent.change(nameInput, {
        target: {
            value: mockPoll.creator.name
        }
    })
    fireEvent.click(getByTestId('submit-button'));
    expect(getByTestId('loading-state'));
    await waitForElement(() => getByTestId('participant-error'));
    fireEvent.click(getByTestId('back-button'));
})