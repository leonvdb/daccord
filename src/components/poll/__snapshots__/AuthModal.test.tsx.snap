// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthModal /> to show Modal on Click 1`] = `
<div
  class="modal-dialog"
  data-testid="auth-modal"
  placement="right"
  role="document"
  target="Modal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <h5
        class="modal-title"
      >
        Become a participant of "
        Test Poll
        "
      </h5>
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          aria-hidden="true"
        >
          Ã—
        </span>
      </button>
    </div>
    <div
      class="modal-body"
    >
      <form>
        <div
          class=" sc-kkGfuU bZZBhd"
        >
          <h5
            class="sc-chPdSV jdQlEK"
          >
            Name
          </h5>
          <input
            class=""
            data-testid="name-input"
            name="name"
            placeholder="Enter Name"
            type="text"
            value=""
          />
        </div>
        <div
          class=" sc-kkGfuU bZZBhd"
        >
          <h5
            class="sc-chPdSV jdQlEK"
          >
            Email
          </h5>
          <input
            class=""
            data-testid="email-input"
            name="email"
            placeholder="Enter Email"
            type="text"
            value=""
          />
        </div>
        <button
          class="btn btn-secondary mx-auto btn-block w-50 mt-5"
          data-testid="submit-button"
          type="submit"
        >
          Continue
        </button>
      </form>
      <p
        class="text-center my-2"
      >
        or
      </p>
      <button
        class="btn btn-outline-info btn-block w-50 mx-auto"
      >
        Sign in
      </button>
    </div>
  </div>
</div>
`;
